hubspot3/tickets.py

```diff
@@ -78,3 +90,185 @@ def get_all(self, limit: int = -1, **options) -> list:
+        # default properties to fetch
+        properties = set(self.default_batch_properties)
+
+        # append extras if they exist
+        if extra_properties:
+            if isinstance(extra_properties, list):
+                properties.update(extra_properties)
+            if isinstance(extra_properties, str):
+                properties.add(extra_properties)
+        if with_history:
+            property_name = "propertiesWithHistory"
+        else:
+            property_name = "properties"
+        params = {}
+        params[property_name] = list(properties)
+        params['includeDeletes'] = True
+        # run the ids as a list of 100
+        batch = {}
+        while len(ids) > 0:
+            partial_ids = ids[:100]
+            ids = ids[100:]
+            partial_batch = self._call(
+                "objects/tickets/batch-read",
+                method="POST",
+                doseq=True,
+                params=params,
+                data={"ids": partial_ids}
+            )
+            batch.update(partial_batch)
+        return batch
+
+    def get_batch(self, ids: List[int],
+                  extra_properties: Union[List[str], str] = None) -> List[dict]:
+        """given a batch of ticket, get more of their info"""
+        batch = self._get_batch(ids, extra_properties=extra_properties, with_history=False)
+        return [prettify(batch[ticket_id], id_key="objectId") for ticket_id in batch]
+
+    def get_batch_with_history(self, ids: List[int],
+                               extra_properties: Union[List[str], str] = None) -> Dict[str, dict]:
+        """given a batch of ticket, get more of their info with history"""
+        batch = self._get_batch(ids, extra_properties=extra_properties, with_history=True)
+        return batch
+
+    def _get_recent(
+        self,
+        recency_type: str,
+        limit: int = -1,
+        ticket_id_offset: int = 0,
+        time_offset: int = 0,
+        max_time_diff_ms: int = 60000,
+        **options
+    ) -> List:
+
+        limited = limit > 0
+        total_tickets = 0
+        finished = False
+        while not finished:
+            params = {}
+            if ticket_id_offset:
+                params["objectId"] = ticket_id_offset
+            if time_offset:
+                params["timestamp"] = int(time_offset)
+            params["changeType"] = recency_type
+            changes = self._call(
+                "change-log/tickets",
+                method="GET",
+                params=params,
+                doseq=True,
+                **options
+            )
+            total_tickets += len(changes)
+            finished = (len(changes) == 0) or (limited and total_tickets >= limit)
+            if len(changes) > 0:
+                ticket_id_offset = changes[-1]["objectId"]
+                time_offset = changes[-1]["timestamp"]
+
+                ids = set([change["objectId"] for change in changes])
```

Comment:
This line can be written as `ids = set(change["objectId"] for change in changes)` to avoid converting to a list. If the project is python3 only (it has typing, so probably?) then there's the set syntax-sugar: `ids = {change["objectId"] for change in changes}`.

Reply:
thanks :) didn't know that